### DISCUSSION TITLE
Alpaca-llama what works and doesn't for my setup.

### DISCUSSION THREAD

**Discussion by thistleknot** [oobabooga/text-generation-webui#1278] [Q&A]
#ozcur

`python server.py --model alpaca-native-4bit --wbits 4 --groupsize 128 --listen --auto-devices --gpu-memory 3500MiB`

results in the infamous

`
File "/mnt/distvol/text-generation-webui/repositories/GPTQ-for-LLaMa/quant.py", line 427, in forward
    quant_cuda.vecquant4matmul(x, self.qweight, y, self.scales, self.qzeros, self.groupsize)
AttributeError: module 'quant_cuda' has no attribute 'vecquant4matmul'
`

However, I'm able to run
`python server.py --model decapoda-research_llama-7b-hf --lora alpaca-lora-7b --auto-devices --listen
`
but not
`python server.py --num_gpus=1 --model decapoda-research_llama-7b-hf --lora alpaca-lora-7b --gptq-bits 4 --gptq-pre-layer 20`

deepspeed throws an error about llamatokenizer

I'm able to run in python
```
from huggingface_hub import hf_hub_download
from pyllamacpp.model import Model

#Download the model
hf_hub_download(repo_id="LLukas22/alpaca-native-7B-4bit-ggjt", filename="ggjt-model.bin", local_dir=".")

#Load the model
model = Model(ggml_model="ggjt-model.bin", n_ctx=2000)

#Generate
prompt="The meaning of life is"

result=model.generate(prompt,n_predict=50)
```

but I cannot run 
`python server.py --model alpaca-native-7B-4bit-ggjt --wbits 4 --pre_layer 20`

error: "Could not find the quantized model in .pt or .safetensors format, exiting"

this worked (and is the best performing)
```
git clone https://huggingface.co/Sosaka/Alpaca-native-4bit-ggml/tree/main
cd Alpaca-native-4bit-ggml
git lfs pull
cd ..
mv Alpaca-native-4bit-ggml models ggml-alpaca-7b-q4
pip install llama-cpp-python

python server.py --model ggml-alpaca-7b-q4 --listen --auto-devices
```

**Comment by MaiJames** [comment]
Hi! I think you are mixing a couple of things here.
There are full size models, that can be loaded at **full precision**, or in **8 bit**.
And there are quantized models, that can only be loaded in the quantization bits that the model has (**4 bit**, **3 bit**). If the model has been quantized with a group size, you have to also match this when loading the model.
The lower the bits, the lower the quality, but you can run it on lower specs and you get faster inference speed.

That said, on your line:
`python server.py --model decapoda-research_llama-7b-hf --lora alpaca-lora-7b --auto-devices --listen`

I see no problems:
- You load the 7B model with full precision, with a 7B lora.
- Auto-devices splits the model across the GPU & CPU.
- Listen makes the UI reachable on local network.

On the second:
`python server.py --num_gpus=1 --model decapoda-research_llama-7b-hf --lora alpaca-lora-7b --gptq-bits 4 --gptq-pre-layer 20`

- You are using arguiment flags that don't exist in the text-generation-webui AFAIK: `--num_gpus=1 --gptq-bits 4 --gptq-pre-layer 20`. You have a list of the [available flags on the main page](https://github.com/oobabooga/text-generation-webui#basic-settings).
- The flag `--gptq-bits 4` suggests that you are trying to load a full precision model in 4 bit, and as I said before, it is not possible.
- There are not quantized LoRas or compatible LoRas with quantized models yet I think, so trying to load a LoRa with a 4 bit model would also give an error.

On the third example that you give:
```
from huggingface_hub import hf_hub_download
from pyllamacpp.model import Model

#Download the model
hf_hub_download(repo_id="LLukas22/alpaca-native-7B-4bit-ggjt", filename="ggjt-model.bin", local_dir=".")

#Load the model
model = Model(ggml_model="ggjt-model.bin", n_ctx=2000)

#Generate
prompt="The meaning of life is"

result=model.generate(prompt,n_predict=50)
```

You are not running the text-generation-webui, and are loading a **pyllama.cpp model**.
These models are in a **GGJT format**, and are not compatible with the text generation webui for now.
 
So, when you do this:
`python server.py --model alpaca-native-7B-4bit-ggjt --wbits 4 --pre_layer 20`

You are trying to load a model that is not compatible, so you get an error.
To run models on the text-generation-webui, you have to look for the models without GGJT (pyllama.cpp format) or GGML (alpaca.cpp format), although compatibility with GGML format was added.

So for example, instead of the "LLukas22/alpaca-native-7B-4bit-ggjt", if you go to the repo, you see that

> Converted version of [Sosaka/Alpaca-native-4bit-ggml](https://huggingface.co/Sosaka/Alpaca-native-4bit-ggml) with ggjt magic for use in [llama.cpp](https://github.com/ggerganov/llama.cpp) or [pyllamacpp](https://github.com/nomic-ai/pyllamacpp).

If you go to the mentioned source "Sosaka/Alpaca-native-4bit-ggml":

> This is a https://huggingface.co/chavinlo/alpaca-native converted in OLD GGML (alpaca.cpp) format and quantized to 4 bits to run on CPU with 5GB of RAM.

You see that the original model, that works with the text-generation-webui is this one: [https://huggingface.co/chavinlo/alpaca-native](https://huggingface.co/chavinlo/alpaca-native)


I recomment you have a look a the [flags](https://github.com/oobabooga/text-generation-webui#basic-settings) and find the best that suit your current specs.

For example, I see you are using the `--pre_layer` one, and this one allocates layers of the model on the GPU, so the bigger the number, the more GPU you use (I think each layer was around 250MB, not sure about this). The rest of the model would run on CPU. The inference speed on CPU is way slower than on GPU, so the more layers you can allocate on the GPU without getting an error, the faster it would be.

I hope it helps.

**Comment by thistleknot** [comment]
The end goal isn't so much trying to find which model works and parameter's, as I found it through trial and error (I became aware of the quantize flags needing to be applied to a quantized model).

What I'm trying to do (maybe I should have stated this before hand).  Is **trying to run a llama alpaca model that was quantized to fit in 4GB of Ram (or 5GB?) so I can put it on my GPU with hopefully deepspeed**, if deepspeed doesn't work, then I'd love it if it would still get on the gpu using just python server.py, but the above was my many attempts at trying just models, and which one worked.  Thanks for the clarification on the alpaca.cpp model (that I wasn't sure what would work, and what wouldn't), but the issue still remains that I'm using an alpaca llama model 4B quantized, and it's all on the CPU.

Is there a llama-alpaca model that runs on the GPU (that I might be able to run with deepspeed)?

It sounds like, from what you are telling me, if I wanted, I could try the 
https://huggingface.co/chavinlo/alpaca-native

but is it possible to quantize this on the fly?  Or does that need to be done beforehand (I assumed this is the case, that you cannot ad-hoc quantize a model from a larger to a lower at run time, that you need to preprocess the model in it's entirety and output a new model, similar to the alpaca-native-7B-4bit-ggjt model).  I haven't had success with loading anything on a 4GB VRAM card (even with deepspeed) > 4B parameter's... and I believe this is 13B?

I usually look at the size of the model as an estimate of it's # of parameter's... for example.  GPT Neo 2.7B is 30GB, and this alpaca-native is also about 30GB, but this alpaca-native is supposed to be 13B... so maybe that assumption using GB is off.

That alpaca-native is too large to fit in my setup...

But it looks like a great base for me to start quantizing from (I setup a paperspace account so I can get access to larger gpu's than my local setup allows).

**The end goal is a 4bit quantized alpaca-native model that fits in a gpu for use with this text-generation-webui.**
